### ISSUE TITLE
Should MVar constructors return `Task`s?

### ISSUE THREAD

**Issue by oleg-py** [monix/monix#535]
Monix MVar constructors are not referentially transparent:

```scala
def mVar = MVar.empty[Int]
mVar.put(0).flatMap(_ => mVar.put(0)) // <-> Task.unit

// vs

val mVar = MVar.empty[Int]
mVar.put(0).flatMap(_ => mVar.put(0)) // <-> Task.never
```

[Haskell](https://hackage.haskell.org/package/base-4.0.0.0/docs/Control-Concurrent-MVar.html#v%3AnewMVar) and [Scalaz](https://github.com/scalaz/scalaz/blob/series/7.3.x/concurrent/src/main/scala/scalaz/concurrent/MVar.scala#L31) constructors both return `IO[MVar[A]]`, since allocating a mutable cell is an effect.

Is it intentional that Monix MVar does not wrap creation in `Task`?

**Comment by alexandru** [comment]
You are right, didn't think that mattered when I designed the API, it's a slip-up 🤕

Are you interested in doing a PR?

**Comment by oleg-py** [comment]
Sure. I'll take a look tomorrow

**Comment by Avasil** [comment]
How about ```Sync```?

**Comment by alexandru** [comment]
@Avasil I've replied to your question on that PR. 

@oleg-py Just a note — I haven't been paying attention to this fact and we probably have other abstractions that need this.

`TaskLocal` definitely needs it and changing this one doesn't hurt so much because it wasn't available in Monix 2.3.x.

I also want your opinion on:

1. `TaskSemaphore`
2. `TaskCircuitBreaker`

These also represent shared state that might violate referential transparency under certain conditions, however I would not change them.

`TaskSemaphore` - is arguably OK in spite of the shared state, because all it does is to delay tasks until slots are available; and I also think that if we box this into a `Task` its usage gets more difficult because you can no longer have a global and people need a global since the main purpose is to protect a data source.

`TaskCircuitBreaker` - again, the same mention - users actually need shared state between tasks, as the goal is to protect a data source that's having issues and boxing it into the `Task` context makes its usage more difficult.

Well, I guess in all cases users can just use `memoize` on the shared `Task` reference if they want to. But that seems awkward and if for the general case people end up with memoization anyway, then I'd rather not do it, especially since it's a breaking change.

So what do you think of these?

**Comment by alexandru** [comment]
Thinking about it and discussing it on the `cats-effect` channel, I think it's pretty much settled, we have to do it for all.

Just one mention — we have to be very clear in ScalaDoc about the options that users have. `Task` can do `.memoize` and this can help users in sharing global references. I propose a mention of `.memoize` for every one of these builders.

I've created a new issue for it: https://github.com/monix/monix/issues/539
